File: typescript/src/replay.ts
import { GameSession, GameVersion } from "./session.js";
import type { GameSessionRef } from "./identity.js";
import type { TransactionalRandomSource } from "./randomness.js";
import type {
  AppliedTransition,
  GameAction,
  GameRules,
  GameSnapshot,
  GameStateOwnership,
} from "./session.js";

export class ReplayResult<State, Outcome, Event> {
  public readonly transitions: readonly AppliedTransition<Event>[];

  public constructor(
    public readonly session: GameSession<State, Outcome>,
    transitions: readonly AppliedTransition<Event>[],
  ) {
    this.transitions = Object.freeze([...transitions]);
    Object.freeze(this);
  }
}

export function replay<State, Action, Event, Outcome, Rejection, Checkpoint>(
  reference: GameSessionRef,
  initialState: State,
  ownership: GameStateOwnership<State, Outcome>,
  rules: GameRules<State, Action, Event, Outcome, Rejection>,
  actions: Iterable<GameAction<Action>>,
  random: TransactionalRandomSource<Checkpoint>,
): ReplayResult<State, Outcome, Event> {
  const session = GameSession.create<State, Outcome>(
    reference,
    initialState,
    ownership,
  );
  session.start(GameVersion.ZERO);
  return applyRecordedActions(session, rules, actions, random);
}

export function replayFromSnapshot<
  State,
  Action,
  Event,
  Outcome,
  Rejection,
  Checkpoint,
>(
  snapshot: GameSnapshot<State, Outcome>,
  ownership: GameStateOwnership<State, Outcome>,
  rules: GameRules<State, Action, Event, Outcome, Rejection>,
  actions: Iterable<GameAction<Action>>,
  random: TransactionalRandomSource<Checkpoint>,
): ReplayResult<State, Outcome, Event> {
  return applyRecordedActions(
    GameSession.restore(snapshot, ownership),
    rules,
    actions,
    random,
  );
}

function applyRecordedActions<
  State,
  Action,
  Event,
  Outcome,
  Rejection,
  Checkpoint,
>(
  session: GameSession<State, Outcome>,
  rules: GameRules<State, Action, Event, Outcome, Rejection>,
  actions: Iterable<GameAction<Action>>,
  random: TransactionalRandomSource<Checkpoint>,
): ReplayResult<State, Outcome, Event> {
  const transitions: AppliedTransition<Event>[] = [];
  for (const action of actions) {
    transitions.push(session.apply(rules, action, random));
  }
  return new ReplayResult(session, transitions);
}

File: rust/src/replay.rs
use crate::{
    AppliedTransition, GameAction, GameExecutionError, GameRules, GameSession, GameSessionRef,
    GameSnapshot, GameVersion, RandomSource,
};

/// Reconstructed authoritative state and the transitions generated while replaying.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ReplayResult<S, O, E> {
    session: GameSession<S, O>,
    transitions: Vec<AppliedTransition<E>>,
}

/// Result type shared by initial-state and snapshot replay operations.
pub type ReplayExecutionResult<R> = Result<
    ReplayResult<<R as GameRules>::State, <R as GameRules>::Outcome, <R as GameRules>::Event>,
    GameExecutionError<<R as GameRules>::Error>,
>;

impl<S, O, E> ReplayResult<S, O, E> {
    #[must_use]
    pub const fn session(&self) -> &GameSession<S, O> {
        &self.session
    }

    #[must_use]
    pub fn transitions(&self) -> &[AppliedTransition<E>] {
        &self.transitions
    }

    #[must_use]
    pub fn into_parts(self) -> (GameSession<S, O>, Vec<AppliedTransition<E>>) {
        (self.session, self.transitions)
    }
}

/// Reconstructs a session from its initial state and ordered accepted actions.
///
/// The core starts the session at version zero, then enforces every recorded
/// action's version and lifecycle precondition.
///
/// # Errors
///
/// Returns the first lifecycle, version, or domain-rule failure.
pub fn replay<R, I>(
    reference: GameSessionRef,
    initial_state: R::State,
    rules: &R,
    actions: I,
    random: &mut dyn RandomSource,
) -> ReplayExecutionResult<R>
where
    R: GameRules,
    I: IntoIterator<Item = GameAction<R::Action>>,
{
    let mut session = GameSession::create(reference, initial_state);
    session.start(GameVersion::ZERO)?;
    apply_recorded_actions(session, rules, actions, random)
}

/// Continues replay from a validated snapshot.
///
/// # Errors
///
/// Returns an invalid-snapshot error or the first action failure.
pub fn replay_from_snapshot<R, I>(
    snapshot: GameSnapshot<R::State, R::Outcome>,
    rules: &R,
    actions: I,
    random: &mut dyn RandomSource,
) -> ReplayExecutionResult<R>
where
    R: GameRules,
    I: IntoIterator<Item = GameAction<R::Action>>,
{
    let session = GameSession::restore(snapshot)?;
    apply_recorded_actions(session, rules, actions, random)
}

fn apply_recorded_actions<R, I>(
    mut session: GameSession<R::State, R::Outcome>,
    rules: &R,
    actions: I,
    random: &mut dyn RandomSource,
) -> ReplayExecutionResult<R>
where
    R: GameRules,
    I: IntoIterator<Item = GameAction<R::Action>>,
{
    let mut transitions = Vec::new();
    for action in actions {
        transitions.push(session.apply(rules, &action, random)?);
    }
    Ok(ReplayResult {
        session,
        transitions,
    })
}
